File: kotlin/src/main/kotlin/org/katas/Birthdays.kt
package org.katas

import java.util.*
import java.util.stream.Collectors

class Birthday(val date: Date, val name: String)


class BirthdayNotebook {
    private val birthdays: MutableMap<String, MutableList<Birthday>> = HashMap()
    fun addBirthday(name: String?, date: Date): BirthdayNotebook {
        val dateAsStr = date.month.toString() + "/" + date.date
        var birthdaysOnDate = birthdays[dateAsStr]
        if (birthdaysOnDate == null) {
            birthdaysOnDate = ArrayList()
            birthdays[dateAsStr] = birthdaysOnDate
        }
        birthdaysOnDate.add(Birthday(date, name!!))
        return this
    }

    fun getBirthdays(date: Date): List<String> {
        val dateAsStr = date.month.toString() + "/" + date.date
        val birthdaysOnDate = birthdays[dateAsStr]
            ?: return emptyList()
        return birthdaysOnDate.stream().map { b: Birthday -> b.name }.collect(Collectors.toList())
    }
}

File: php/src/Birthdays.php
<?php
class BirthDay {
  public $date;
  public $name;

  public function __construct($date, $name) {
    $this->date = $date;
    $this->name = $name;
  }
}

class BirthdayNotebook {
  private $birthdays = array();

  public function addBirthday($name, $date) {
    $dateAsStr = $date->format('n/j');
    $this->birthdays[$dateAsStr][] = new Birthday($date, $name);
    return $this;
  }

  public function getBirthdays($date) {
    $dateAsStr = $date->format('n/j');
    return array_map(function($b) { return $b->name; }, $this->birthdays[$dateAsStr] ?? []);
  }
}
